### PR TITLE
Add type hints to atom.py

### PR DESCRIPTION
**Pull request by pyssling**
This adds type hints to atom.py, or silences errors by mypy by adding suppressions. There are substantial opportunities for further improvement here, though that probably means refactoring the actual code.

### CODE REVIEW

File: tables/atom.py [hunk 1]
```diff
@@ -652,29 +675,29 @@ def _create_numeric_class(baseclass, itemsize):
...
 
-Int8Atom = _create_numeric_class(IntAtom, 1)
-Int16Atom = _create_numeric_class(IntAtom, 2)
-Int32Atom = _create_numeric_class(IntAtom, 4)
-Int64Atom = _create_numeric_class(IntAtom, 8)
-UInt8Atom = _create_numeric_class(UIntAtom, 1)
-UInt16Atom = _create_numeric_class(UIntAtom, 2)
-UInt32Atom = _create_numeric_class(UIntAtom, 4)
-UInt64Atom = _create_numeric_class(UIntAtom, 8)
+Int8Atom : type[IntAtom] = _create_numeric_class(IntAtom, 1)
```
Comment (pyssling) [inline tables/atom.py:684]: I'm not especially fond of this, and I'm not sure that the typing is 'correct'. An option would be to simply expand this and create the classes vebatim. I think that would probably be more elegant for the type system, but would of course require registering them in some other way.

**Review by avalentino — COMMENTED** [review]
Dear @pyssling I really appreciate and support any activity targeted to a modernization of the PyTables codebase. A huge thank you from my side for submitting this PR.
Your PR looks very nice to me.
The only problem is that you are using constructs and functions that have been introduced in python after v3.9 which is the older Python version that currently PyTables supports.
My request would be to use [typing-extensions](https://pypi.org/project/typing-extensions) to ensure compatibility with older Python versions.
If necessary, you could bump the minimum Python version in this PR (and update the CI configuration accordingly) but I would avoid to reduce to much the number of supported Python versions.

**Comment by pyssling** [comment]
> My request would be to use [typing-extensions](https://pypi.org/project/typing-extensions) to ensure compatibility with older Python versions.
> If necessary, you could bump the minimum Python version in this PR (and update the CI configuration accordingly) but I would avoid to reduce to much the number of supported Python versions.

I'll try to move to using [typing-extension]. Thank you. I was expecting substantial revision on this PR.

About the `_create_numeric_class()` and similarly created classes, do you think we should expand them or is this undesirable? Expanding them would generate more code, but would simplify 'correct' typing.

**Comment by avalentino** [comment]
> About the `_create_numeric_class()` and similarly created classes, do you think we should expand them or is this undesirable? Expanding them would generate more code, but would simplify 'correct' typing.

I'm fine with expanding  `_create_numeric_class()`

**Comment by pyssling** [comment]
This may still need some work for python 3.9 and other matrix builds, but please review @avalentino .

File: tables/atom.py [hunk 2]
```diff
@@ -236,57 +225,75 @@ class Atom(metaclass=MetaAtom):
 
     .. rubric:: Atom attributes
 
-    .. attribute:: dflt
+    .. autoattribute:: dflt
```
Comment (avalentino) [inline tables/atom.py:228]: General comment:
I think that using `..auto*` primitives for attributes has the drawback that the attribute documentation is not available and visible on the interactive help system.
I'm not sure that we should go in that direction honestly, but I'm open to discuss.

File: tables/atom.py [hunk 3]
```diff
@@ -577,104 +589,202 @@ def _is_equal_to_atom(self, atom):
...
     """Defines an atom of a signed integral type (int kind)."""
 
-    kind = 'int'
-    signed = True
+    kind: str = 'int'
+    signed: bool = True
     _deftype = 'int32'
     _defvalue = 0
-    __init__ = _abstract_atom_init(_deftype, _defvalue)
+    __init__  = _abstract_atom_init(_deftype, _defvalue) # type: ignore[assignment]
```
Comment (avalentino) [inline tables/atom.py:652]: extra white space before "="

File: tables/atom.py [hunk 4]
```diff
@@ -577,104 +589,202 @@ def _is_equal_to_atom(self, atom):
...
     """Defines an atom of an unsigned integral type (uint kind)."""
 
-    kind = 'uint'
-    signed = False
+    kind: str = 'uint'
+    signed: bool = False
     _deftype = 'uint32'
     _defvalue = 0
-    __init__ = _abstract_atom_init(_deftype, _defvalue)
+    __init__  = _abstract_atom_init(_deftype, _defvalue) # type: ignore[assignment]
```
Comment (avalentino) [inline tables/atom.py:662]: Extra white space

File: tables/atom.py [hunk 5]
```diff
@@ -1078,12 +1192,9 @@ class VLStringAtom(_BufferedAtom):
     type = 'vlstring'
     base = UInt8Atom()
 
-    def _tobuffer(self, object_):
-        if isinstance(object_, str):
-            warnings.warn("Storing non bytestrings in VLStringAtom is "
-                          "deprecated.", DeprecationWarning)
```
Comment (avalentino) [inline tables/atom.py:1084]: We should have an entry in the release notes for this change.

File: tables/atom.py [hunk 6]
```diff
@@ -1120,25 +1231,19 @@ class VLUnicodeAtom(_BufferedAtom):
     # NumPy scalars inherit the UCS-2 encoding from Python (see
     # NumPy ticket #525).  Since ``_tobuffer()`` can't return an
     # array, we must override ``toarray()`` itself.
-    def toarray(self, object_):
-        if isinstance(object_, bytes):
-            warnings.warn("Storing bytestrings in VLUnicodeAtom is "
-                          "deprecated.", DeprecationWarning)
```
Comment (avalentino) [inline tables/atom.py:1126]: Same here

File: tables/atom.py [hunk 7]
```diff
@@ -1120,25 +1231,19 @@ class VLUnicodeAtom(_BufferedAtom):
...
             buffer=uarr, dtype=self.base.dtype, shape=len(ustr))
 
-    def _tobuffer(self, object_):
+    def _tobuffer(self, object_: str):
         # This works (and is used) only with UCS-4 builds of Python,
         # where the width of the internal representation of a
         # character matches that of the base atoms.
-        if isinstance(object_, bytes):
-            warnings.warn("Storing bytestrings in VLUnicodeAtom is "
-                          "deprecated.", DeprecationWarning)
```
Comment (avalentino) [inline tables/atom.py:1140]: item

**Review by avalentino — CHANGES_REQUESTED** [review]
The PR looks very good at this stage.
There are few small changes do be addressed )see comments for details).
Thanks @pyssling

File: tables/atom.py [hunk 8]
```diff
@@ -149,15 +134,16 @@ def _cmp_dispatcher(other_method_name):
     This is part of the implementation of a double dispatch pattern.
     """
 
-    def dispatched_cmp(self, other):
+    def dispatched_cmp(self: object, other: object) -> bool:
```
[the remark this answers is not in the document]
Comment (pyssling) [reply tables/atom.py:137]: Removing it then, it doesn't affect type checkers as it's only an internal definition.

File: tables/atom.py (hunk 2 above)
Comment (pyssling) [reply tables/atom.py:228]: Undoing then, this has little to do with the type hints anyway.

File: tables/atom.py (hunk 3 above)
Comment (pyssling) [reply tables/atom.py:652]: done.

File: tables/atom.py (hunk 4 above)
Comment (pyssling) [reply tables/atom.py:662]: done.

File: tables/atom.py (hunk 5 above)
Comment (pyssling) [reply tables/atom.py:1084]: Added a small section in the release notes.

File: tables/atom.py (hunk 6 above)
Comment (pyssling) [reply tables/atom.py:1126]: Done.

File: tables/atom.py (hunk 7 above)
Comment (pyssling) [reply tables/atom.py:1140]: Sorry, not following this comment.

Comment (avalentino) [reply tables/atom.py:1140]: Also this change should be mentioned in the release notes.

Comment (avalentino) [reply tables/atom.py:1140]: OK it is already mentioned together in the end in the release note item you have already added.

**Comment by avalentino** [comment]
There seems to be a small issue with the "CI/Sdist with cblocs2" job.
It is apparently an issue with the import of "Any".
Please note that, according to the [typing_extensions documentation](https://typing-extensions.readthedocs.io/en/latest/#Any), "New in version 4.4.0: Added to support inheritance from Any."
Hope this helps.

**Comment by pyssling** [comment]
@avalentino I fixed one of the pipelines, found another place I need to update the requirements. However, there is still one remaining build failure, on 'ubuntu-latest 3.12 numpy>=2.dev0' . I can't see that this one has to do with any of my changes though?

**Review by avalentino — APPROVED** [review]

**Comment by avalentino** [comment]
Thanks @pyssling

**Pull request merged**